### PR TITLE
Fix inconsistent requirement for kind in strict json decoding

### PR DESCRIPTION
**Pull request by liggitt**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

When using non-strict unstructured json decoding, kind and apiVersion are required.

When using strict unstructured json decoding, missing kind and apiVersion were tolerated. This adds tests and makes strict decoding consistent.

```release-note
Custom resource requests with fieldValidation=Strict consistently require apiVersion and kind, matching non-strict requests
```

/cc @apelisse @kevindelgado 
/sig api-machinery

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json_test.go [hunk 1]
```diff
@@ -124,13 +124,290 @@ func TestDecode(t *testing.T) {
...
-	}{
+	}
+
+	testCases := []testCase{
+		// missing metadata without into, typed creater
+		{
+			data: []byte("{}"),
+
+			expectedGVK: &schema.GroupVersionKind{},
+			errFn:       func(err error) bool { return strings.Contains(err.Error(), "Object 'Kind' is missing in") },
```
Comment (apelisse) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json_test.go:135]: nit but, should we report both missing kind and version at the same time?

**Review by apelisse — COMMENTED** [review]
looks good

**Comment by apelisse** [comment]
/lgtm
I don't think we'll want to do something about my nit, but in case:
/hold

**Comment by liggitt** [comment]
/hold

hmmm, will dig on those. I verified kind and apiVersion were required when making custom resource requests, but will look at these other uses.

**Comment by liggitt** [comment]
/hold cancel

the implications of adding the apiVersion requirement to existing decoding paths are bigger than we should tackle here. made strict consistent with non-strict for now, and opened #109019 to track

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json_test.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json_test.go:135]: not changing existing non-strict behavior in this PR

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go [hunk 2]
```diff
@@ -340,6 +340,7 @@ func (s unstructuredJSONScheme) Decode(data []byte, _ *schema.GroupVersionKind,
 	if len(gvk.Kind) == 0 {
 		return nil, &gvk, runtime.NewMissingKindErr(string(data))
 	}
+	// TODO(109019): require apiVersion here as well
```
Comment (apelisse) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go:343]: This refers to this pull-request, is this expected?

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go]: sigh

**Review by apelisse — COMMENTED** [review]
Thanks
/lgtm

**Pull request merged**